File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let slowNode = head;
    let fastNode = head;
    let prevNode = null;

    for(let nodeIdx = 0; nodeIdx < n; nodeIdx++){
        fastNode = fastNode.next;
    }

    while(fastNode){
        prevNode = slowNode;
        slowNode = slowNode.next;
        fastNode = fastNode.next;
    }

    if(prevNode === null) return head.next;

    prevNode.next = slowNode.next;
    slowNode.next = null;

    return head;
    /** 
    if(n === 1 && head.next === null) return null;

    let linkedListLen = 0;
    let currentNode = head;

    while(currentNode){
        currentNode = currentNode.next;
        linkedListLen += 1;
    }

    let deletingNodeIdx = linkedListLen - n;
    let prevNode = head;
    currentNode = head;

    // console.log(deletingNodeIdx, linkedListLen);

    for(let nodeIdx = 0; nodeIdx < deletingNodeIdx; nodeIdx++){
        prevNode = currentNode;
        currentNode = currentNode.next;
    }

    prevNode.next = currentNode.next;
    currentNode.next = null;

    return head;
    */
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # Logic: Find (k-1)th node and then reset the
        # pointers to remove the kth node;
        if n <= 0 or head == None:
            return head
        
        if head.next == None and n == 1:
            return head.next
        
        slow_node = head
        fast_node = head
        prev_node = None

        # move fast_node by n - 1 step ahead in the list;
        for node_idx in range(n):
            if fast_node:
                fast_node = fast_node.next
        
        if fast_node == None:
            return head.next
        
        while fast_node:
            prev_node = slow_node
            slow_node = slow_node.next
            fast_node = fast_node.next

        prev_node.next = slow_node.next
        slow_node.next = None

        return head